src/server/utils.ts

```diff
@@ -224,3 +229,185 @@ export function compareTextEdits(a: TextEdit, b: TextEdit): number {
     if (a.newText > b.newText) return 1;
     return compareRanges(a.range, b.range);
 }
+
+type Version = string;
+
+export function extractMocVertion(filename: string): Version | undefined {
+    const versionRegex = /^moc-(.+)\.js$/;
+    const res = filename.match(versionRegex);
+    if (res && res[1]) {
+        return res[1];
+    }
+    return;
+}
+
+/**
+ * Attempts to determine the Motoko compiler (moc) version used in the project.
+ *
+ * The function first tries to get the moc version from mops toolchain.
+ * If that fails, it falls back to using the moc binary from the dfx cache.
+ *
+ * @returns {ResultAsync<Version, Error>} - The detected moc version or an error.
+ */
+export function getWorkspaceMocVersion(
+    workspaceDir: string,
+): ResultAsync<Version, Error> {
+    function getMocVersion(mocPath: string): ResultAsync<Version, Error> {
+        return ResultAsync.fromThrowable(
+            () => execa(mocPath, ['--version']),
+            (e) => (e instanceof Error ? e : new Error(String(e))),
```

Comment:
Why do we perform this conversion to `Error`? Wouldn't it be better to return `any` or `unknown`?

Or does the callee of these functions only handle `Error`?

Reply:
Ok, I have no strong opinions about it, and this pattern seems common in `vscode-motoko`, so let's leave it as it is.